sklearn/linear_model/ridge.py

```diff
@@ -367,11 +367,29 @@ def _ridge_regression(X, y, alpha, sample_weight=None, solver='auto',
                       return_n_iter=False, return_intercept=False,
                       X_scale=None, X_offset=None):
 
-    if return_intercept and sparse.issparse(X) and solver != 'sag':
-        if solver != 'auto':
-            warnings.warn("In Ridge, only 'sag' solver can currently fit the "
-                          "intercept when X is sparse. Solver has been "
-                          "automatically changed into 'sag'.")
+    has_sw = sample_weight is not None
+
+    def _select_auto_mode():
+        if return_intercept:
+            # only sag supports fitting intercept directly
+            return "sag"
+        if has_sw:
+            # this should be changed since all solvers support sample_weights
+            return "cholesky"
+        if sparse.issparse(X):
+            return "sparse_cg"
+        return "cholesky"
+
+    if solver == 'auto':
+        solver = _select_auto_mode()
+
+    if solver not in ('sparse_cg', 'cholesky', 'svd', 'lsqr', 'sag', 'saga'):
+        raise ValueError('Solver %s not understood' % solver)
+
+    if return_intercept and solver != 'sag':
+        warnings.warn("In Ridge, only 'sag' solver can currently fit the "
+                      "intercept. Solver has been "
+                      "automatically changed into 'sag'.")
```

Comment:
Is this still true: I though that now ridge supports fitting the intercept via other solvers.

Reply:
Only the `Ridge.fit` method supports fitting intercept with other solvers (i.e. cholesky/sparse_cg). This was already the case before and it should be changed when the `ridge_regression` function is refactored as discussed in #13336.

Reply:
Should we not just raise an error in this case?

Reply:
I think an error is better too.
Doesn't 'saga' also support fitting intercept ?

Reply:
Since #13336 is merged this needs to be update